File: 02-Prompt-Engineering/prompt-engineering-python/examples/1-tipos-de-prompts/0-Role-prompting.py
from langchain_openai import ChatOpenAI
from langchain.prompts import ChatPromptTemplate
from dotenv import load_dotenv
from pydantic import SecretStr
import os
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..'))

from utils.prompt_helpers import print_llm_result

load_dotenv()

system = ("system", 
    """
    Você é um professor universitário de ciência da computação, muito técnico e explica conceitos
    com definições formais e pseudocódigo
    """          
)

system2 = ("system",
    """
    Você é um estudante do ensino médio que está começando a aprender programação.
    Você não é muito técnico e prefere explicar conceitos com palavras e exemplos simples.
    """          
)

user = ("user", "Explique recursão em 50 palavras.")

chat_prompt = ChatPromptTemplate([system, user])
chat_prompt2 = ChatPromptTemplate([system2, user])
messages = chat_prompt.format_messages()

model = ChatOpenAI(
    model="gpt-4o",
    base_url=os.environ["GITHUB_MODELS_ENDPOINT"],
    api_key=SecretStr(os.environ["GITHUB_MODELS_TOKEN"])
)
response = model.invoke(messages)
print_llm_result(str(system), response)

response2 = model.invoke(chat_prompt2.format_messages())
print_llm_result(str(system2), response2)

File: prompt-engineering-typescript/src/examples/1-tipos-de-prompts/0-role-prompting.ts
import { ChatOpenAI } from "@langchain/openai";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import {
  loadEnvironment,
  getApiKey,
  getEndpoint,
  printLlmResult,
} from "../../utils/helpers";

// Carrega as variáveis de ambiente
loadEnvironment();

async function main(): Promise<void> {
  const systemProfessor = `Você é um professor universitário de ciência da computação, muito técnico e explica conceitos
    com definições formais e pseudocódigo.`;

  const systemEstudante = `Você é um estudante do ensino médio que está começando a aprender programação.
    Você não é muito técnico e prefere explicar conceitos com palavras e exemplos simples.`;

  const userMessage = "Explique recursão em 50 palavras.";

  const chat_prompt = ChatPromptTemplate.fromMessages([
    ["system", systemProfessor],
    ["user", userMessage],
  ]);

  const chat_prompt2 = ChatPromptTemplate.fromMessages([
    ["system", systemEstudante],
    ["user", userMessage],
  ]);

  const model = new ChatOpenAI({
    model: "gpt-4o",
    configuration: {
      baseURL: getEndpoint("GITHUB_MODELS_ENDPOINT"),
      apiKey: getApiKey("GITHUB_MODELS_TOKEN"),
    },
  });

  const response = await model.invoke(await chat_prompt.formatMessages({}));
  printLlmResult(systemProfessor, response);

  const response2 = await model.invoke(await chat_prompt2.formatMessages({}));
  printLlmResult(systemEstudante, response2);
}

main();
